### ISSUE TITLE
Crash in [LPVarCache saveDiffs]

### ISSUE THREAD

**Issue by jtrainer** [Leanplum/Leanplum-iOS-SDK#129]
## Steps to Reproduce the Problem

  1. Unknown. This is an infrequent crash reported to Crashlytics that we have not been able to reproduce.

## Specifications

  - Version: LP SDK 2.0.5
  - Platform: iPhone, iOS 11

```
com.apple.main-thread
0  libsystem_kernel.dylib         0x1857b5734 guarded_pwrite_np + 8
1  libsqlite3.dylib               0x186141b00 sqlite3_rekey + 2476
2  libsqlite3.dylib               0x1861045bc sqlite3_free_table + 58480
3  libsqlite3.dylib               0x18614b788 sqlite3_rekey + 42548
4  libsqlite3.dylib               0x1860ffd30 sqlite3_free_table + 39908
5  libsqlite3.dylib               0x186103878 sqlite3_free_table + 55084
6  libsqlite3.dylib               0x186103220 sqlite3_free_table + 53460
7  libsqlite3.dylib               0x1860e0048 sqlite3_step + 21248
8  libsqlite3.dylib               0x1860db7dc sqlite3_step + 2708
9  Leanplum                       0x1018baddc -[LPDatabase runQuery:bindObjects:] + 104
10 Leanplum                       0x1018bc394 +[LPEventDataManager addEvent:] + 204
11 Leanplum                       0x1018af10c -[LeanplumRequest sendEventually] + 428
12 Leanplum                       0x1018ada74 -[LeanplumRequest send] + 32
13 Leanplum                       0x101896ab8 +[Leanplum setUserIdInternal:withAttributes:] + 520
14 Leanplum                       0x101891240 +[Leanplum onStartIssued:] + 108
15 Leanplum                       0x1018967d8 +[Leanplum setUserId:withUserAttributes:] + 684
16 Leanplum                       0x101896508 +[Leanplum setUserId:] + 112
17 --REDACTED--                   0x100efd8ac specialized LeanplumAnalyticsDriver.(start(deviceId : String, onStart : () -> ()) -> ()).(closure #1) (LeanplumAnalyticsDriver.swift:42)
18 --REDACTED--                   0x100efd6a8 partial apply for LeanplumAnalyticsDriver.(start(deviceId : String, onStart : () -> ()) -> ()).(closure #1) (LeanplumAnalyticsDriver.swift)
19 --REDACTED--                   0x100d1b2cc thunk (MainViewController.swift)
20 Leanplum                       0x10188b62c +[Leanplum triggerStartResponse:] + 552
21 Leanplum                       0x10188f644 __59+[Leanplum startWithUserId:userAttributes:responseHandler:]_block_invoke.680 + 112
22 Leanplum                       0x1018bb4f0 -[LPEventCallback invokeError:] + 136
23 Leanplum                       0x1018bc13c +[LPEventCallbackManager invokeErrorCallbacksWithError:] + 232
24 Leanplum                       0x1018aec0c __32-[LeanplumRequest sendRequests:]_block_invoke.172 + 284
25 Leanplum                       0x1018cfd30 __39-[LPNetworkOperation runSynchronously:]_block_invoke_2 + 176
26 libdispatch.dylib              0x18561ea54 _dispatch_call_block_and_release + 24
27 libdispatch.dylib              0x18561ea14 _dispatch_client_callout + 16
28 libdispatch.dylib              0x18562b698 _dispatch_main_queue_callback_4CF$VARIANT$mp + 1016
29 CoreFoundation                 0x185c47344 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
30 CoreFoundation                 0x185c44f20 __CFRunLoopRun + 2012
31 CoreFoundation                 0x185b64c58 CFRunLoopRunSpecific + 436
32 GraphicsServices               0x187a10f84 GSEventRunModal + 100
33 UIKit                          0x18f2bd5c4 UIApplicationMain + 236
34 --REDACTED--                   0x100c79d90 main (AppDelegate.swift:27)
35 libdyld.dylib                  0x18568456c start + 4
```

```
Thread #2
Crashed: NSOperationQueue 0x1c4032620 (QOS: UNSPECIFIED)
0  libobjc.A.dylib                0x184efc430 objc_msgSend + 16
1  Foundation                     0x1865f30f0 _encodeObject + 208
2  Leanplum                       0x1018d6650 +[LPVarCache saveDiffs] + 636
3  Leanplum                       0x1018d6fec +[LPVarCache applyVariableDiffs:messages:updateRules:eventRules:variants:regions:] + 1828
4  Leanplum                       0x1018d62cc +[LPVarCache loadDiffs] + 1132
5  Leanplum                       0x10188f630 __59+[Leanplum startWithUserId:userAttributes:responseHandler:]_block_invoke.680 + 92
6  Leanplum                       0x1018ae6d4 __32-[LeanplumRequest sendRequests:]_block_invoke + 1424
7  Foundation                     0x186675ba0 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16
8  Foundation                     0x1865b5894 -[NSBlockOperation main] + 72
9  Foundation                     0x1865a54c4 -[__NSOperationInternal _start:] + 848
10 Foundation                     0x186677894 __NSOQSchedule_f + 404
11 libdispatch.dylib              0x18561ea14 _dispatch_client_callout + 16
12 libdispatch.dylib              0x185626f08 _dispatch_continuation_pop$VARIANT$mp + 428
13 libdispatch.dylib              0x18562580c _dispatch_async_redirect_invoke$VARIANT$mp + 604
14 libdispatch.dylib              0x18562bcf4 _dispatch_root_queue_drain + 600
15 libdispatch.dylib              0x18562ba38 _dispatch_worker_thread3 + 120
16 libsystem_pthread.dylib        0x1858c706c _pthread_wqthread + 1268
17 libsystem_pthread.dylib        0x1858c6b6c start_wqthread + 4
```

**Comment by alexisoyama** [comment]
Thanks for reporting this. Can you share the full crash report with other threads?

**Comment by jtrainer** [comment]
Full Crash Report [Crash-SaveDiff.txt](https://github.com/Leanplum/Leanplum-iOS-SDK/files/1708598/Crash-SaveDiff.txt)

**Comment by jtrainer** [comment]
Discovered that I misattributed the SDK version and this should have been against version 2.0.4. The crash appears to be fixed in 2.0.5 and we have not seen the crash again. Closing.
